File: appli-web/src/app/designsystem/common_components/progress-bar-primitive.ts
import {
  afterNextRender,
  ChangeDetectionStrategy,
  Component,
  computed,
  input,
  signal,
} from '@angular/core';

/**
 * Primitif de barre de progression — miroir de ProgressBarPrimitive.kt : trough (default bgRecessed)
 * + remplissage couleur, hauteur 7px par défaut, coins 2px. Source unique du look de la barre.
 *
 * Le remplissage est **animé 0 → valeur** : on peint 0 au 1er rendu puis on pose la vraie valeur au
 * paint suivant (`afterNextRender`) → la transition CSS joue le remplissage qui monte. Effet hérité
 * par tous les usages (LabeledProgressBar, etc.). `prefers-reduced-motion: reduce` coupe la transition.
 */
@Component({
  selector: 'app-progress-bar-primitive',
  changeDetection: ChangeDetectionStrategy.OnPush,
  template: `
    <div class="trough" [style.height.px]="height()" [style.background]="troughColor()">
      <div class="fill" [style.width.%]="pct()" [style.background]="color()"></div>
      @if (markerAt() !== null) {
        <div class="marker" [style.left.%]="markerPct()"></div>
      }
    </div>
  `,
  styles: [
    `
      :host {
        display: block;
      }
      .trough {
        position: relative;
        width: 100%;
        border-radius: 2px;
        overflow: hidden;
      }
      /* Repère vertical statique (seuil) : trait fin par-dessus trough ET remplissage. */
      .marker {
        position: absolute;
        top: 0;
        bottom: 0;
        width: 2px;
        transform: translateX(-50%);
        background: var(--app-text-tertiary);
      }
      .fill {
        height: 100%;
        border-radius: 2px;
        /* Remplissage animé 0 → valeur (apparition + changements) via les tokens de motion. */
        transition: width var(--motion-base) var(--motion-ease);
      }
      /* a11y : aucune animation si l'utilisateur a demandé moins de mouvement. */
      @media (prefers-reduced-motion: reduce) {
        .fill {
          transition: none;
        }
      }
    `,
  ],
})
export class ProgressBarPrimitive {
  /** Progression 0..1. */
  readonly progress = input(0);
  readonly color = input('var(--app-primary-action)');
  readonly height = input(7);
  readonly troughColor = input('var(--app-bg-recessed)');
  /** Repère vertical optionnel (0..1, null = aucun) — ex. seuil « idéal » d'une barre de limite
   *  (sucres OMS : marque à 5 % sur la barre bornée au plafond 10 %). Statique, non animé. */
  readonly markerAt = input<number | null>(null);

  /** Faux jusqu'au 1er paint : on rend 0 d'abord, puis la vraie valeur → la transition part de 0. */
  private readonly mounted = signal(false);

  protected readonly pct = computed(
    () => Math.max(0, Math.min(1, this.mounted() ? this.progress() : 0)) * 100,
  );

  protected readonly markerPct = computed(
    () => Math.max(0, Math.min(1, this.markerAt() ?? 0)) * 100,
  );

  constructor() {
    afterNextRender(() => this.mounted.set(true));
  }
}

File: appli-android/app/src/main/java/com/example/sportapp/designsystem/common_components/ProgressBarPrimitive.kt
package com.example.sportapp.designsystem.common_components

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.fillMaxHeight
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import com.example.sportapp.designsystem.theme.appColors

/**
 * Primitif R18 — rend uniquement la barre de progression : trough (couleur de
 * fond, default `appColors.bgRecessed` = thirdBlue) + remplissage couleur,
 * hauteur par défaut 7dp, coins `RoundedCornerShape(2.dp)`. Source unique de
 * vérité du look de la barre, partagée par `LabeledProgressBar`,
 * `RoutineTasksProgressBar`, `PlannedDayProgressBar`.
 *
 * Le `modifier` est appliqué à la trough (extérieur) — c'est là que le caller
 * passe `Modifier.weight(1f)` ou similaire. La `progress` est coerçée 0..1.
 *
 * `troughColor` peut être overridé en `appColors.bgSurface` (boxBlue) quand la
 * bar est rendue sur un container `bgRecessed` (thirdBlue) — sinon trough
 * invisible (même couleur que le container). Cas typique : la bar dans
 * `RoutineTasksProgressBar` qui se posent dans son propre container bgRecessed.
 *
 * `markerAt` : repère vertical optionnel (0..1, null = aucun), trait fin 2dp
 * `textTertiary` par-dessus trough ET remplissage — ex. seuil « idéal » d'une
 * barre de limite (sucres OMS : marque à 5 % sur la barre bornée au plafond
 * 10 %). Miroir du `markerAt` du primitif web.
 */
@Composable
fun ProgressBarPrimitive(
    progress: Float,
    color: Color,
    modifier: Modifier = Modifier,
    height: Dp = 7.dp,
    troughColor: Color = appColors.bgRecessed,
    markerAt: Float? = null,
) {
    Box(
        modifier = modifier
            .height(height)
            .background(troughColor, shape = RoundedCornerShape(2.dp))
    ) {
        Box(
            modifier = Modifier
                .fillMaxHeight()
                .fillMaxWidth(progress.coerceIn(0f, 1f))
                .background(color, shape = RoundedCornerShape(2.dp))
        )
        markerAt?.let { at ->
            Box(
                modifier = Modifier.fillMaxHeight().fillMaxWidth(at.coerceIn(0f, 1f)),
                contentAlignment = Alignment.CenterEnd,
            ) {
                Box(
                    modifier = Modifier
                        .fillMaxHeight()
                        .width(2.dp)
                        .background(appColors.textTertiary)
                )
            }
        }
    }
}
